### PR TITLE
Implement Rejection Sampling Variational Inference

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #63

This implements Rejection Sampling Variational Inference (RSVI) of [Naesseth et al. (2017)](https://arxiv.org/pdf/1610.05683.pdf).

This implements a new class `Rejector` as an abstract base class derived from `Distribution`. The crux to integrating these with Pyro's SVI in `TraceGraph_ELBO` is a new method `Distribution.score_function_term()` that can be overridden to compute `g_cor` in the paper. This method computes a partial score function for a partially reparameterized distribution such that: for an unreparameterized distribution it computes the entire score function, and for a fully reparameterized distribution it computes zero.

## Tasks

- [x] integrate with `Trace_ELBO`
- [x] implement `RejectionGamma` example, e.g. see @slinderman's [notebook](https://github.com/blei-lab/ars-reparameterization/blob/master/gamma/demo.ipynb)
- [x] implement @naesseth's shape augmentation, e.g. see [code](https://github.com/blei-lab/ars-reparameterization)
- [x] write convergence tests
- [x] write smoke tests
- [x] clean up PR to remove irrelevant testing stuff
- [x] fix unit tests

This was joint work by @martinjankowiak and @fritzo

### CODE REVIEW

**Comment by fritzo** [comment]
cc @rachtsingh @naesseth

**Comment by slinderman** [comment]
Glad to see this incorporated into pyro, @fritzo and @martinjankowiak!   One of the points discussed in the paper is that the score function term tends to be higher variance, and in practice it can often be safely ignored at the cost of a small bias.  

You can also ameliorate it to some extent by "shape augmentation," i.e. introducing auxiliary uniform r.v.'s in order to increase the shape parameter and thereby increase the acceptance probability.  I did not implement this in the demo notebook, but we used it in our paper.  The code for that is in this repo: https://github.com/blei-lab/ars-reparameterization and @naesseth can probably provide more details on its implementation.

**Comment by fritzo** [comment]
@martinjankowiak This is almost done, there's just one more test to fix.

Could you please review the changes to `Rejector`? I've simplified and added a `log_scale` argument that is used to correctly compute the `.batch_log_pdf()` scale.

**Comment by martinjankowiak** [comment]
cool awesome. i'll do so in the morning when i have a fresh brain

File: pyro/distributions/distribution.py [hunk 1]
```diff
@@ -187,6 +191,27 @@ def batch_log_pdf(self, x, *args, **kwargs):
...
+        distributions should override this method to compute correct
+        `.score_function` and `.entropy_term` parts.
+
+        :param torch.autograd.Variable x: A single value or batch of values.
+        :return: A `ScoreParts` object containing parts of the ELBO estimator.
+        :rtype: ScoreParts
+        """
+        log_pdf = self.batch_log_pdf(x, *args, **kwargs)
+        if self.reparameterized:
+            return ScoreParts(log_pdf, 0, log_pdf)
```
Comment (martinjankowiak) [inline pyro/distributions/distribution.py]: maybe explicit names in namedtuple here?

**Comment by fritzo** [comment]
@martinjankowiak I believe I've fixed `Rejector` (sorry for the earlier bug!), but `ShapeAugmentedGamma` gradients still appear to be wrong. Do you think it's worth merging this as-is and debugging `ShapeAugmentedGamma` in a follow-up PR that also adds `ShapeAugmentedBeta` etc?

**Comment by fritzo** [comment]
Ready to merge, finally.

**Comment by rmehta1987** [comment]
@fritzo are there any examples of using the accept-reject dirichlet or gamma?

**Pull request merged**